Fix: Map functions crash on out of bounds cases

krishvishal opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

- Closes https://github.com/apache/datafusion/issues/16187

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->
This PR fixes a panic crash in array indexing operations when accessing elements from `map_values()` results that contain structs with `Null`-typed fields.

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->
- `datafusion/functions-nested/src/extract.rs`: Added null field detection and safe handling



## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->
Yes - unit tests are added covering:
- Main crash scenario: Struct with `Null`-typed fields
- Mixed null/valid fields, out-of-bounds access

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->
No, only behavior change. 
<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

comphead commented on datafusion/functions-nested/src/extract.rs:
```diff
@@ -225,6 +225,23 @@ where
+        let has_null_fields = fields.iter().any(|field| field.data_type() == &Null);
+        if has_null_fields {
+            // Instead of trying to extract from malformed struct data and return appropriate nulls
+            let mut null_array_data = Vec::with_capacity(array.len());
```
to return I think we can return NullArray like here?
```
    if values.data_type().is_null() {
         return Ok(Arc::new(NullArray::new(array.len())));
     }
```

comphead commented on datafusion/functions-nested/src/extract.rs:
```diff
@@ -225,6 +225,23 @@ where
     }
 
+    if let DataType::Struct(fields) = values.data_type() {
+        let has_null_fields = fields.iter().any(|field| field.data_type() == &Null);
```
oh, now I realize the problem happens only if the struct has with with `null` as the value

comphead commented:
@krishvishal are you still planning to wrap this PR up?

krishvishal commented:
@comphead I've changed the implementation a bit to handle nulls properly. Previous just outputs `NULL` for queries like `select [named_struct('a', 1, 'b', null)][1];` instead of `{a: 1, b: NULL} `.

I've also added tests to `.array.slt` file. 

Apologies for the delay.

comphead commented:
Thanks @krishvishal the latest version becomes much more complicated compared to prev one. This can be a subject to check the performance.

What is the reason for adding the specific handler, some tests not passing?

krishvishal commented:
@comphead I've had to add the handler because the previous fix caused wrong behavior.

For example the following query currently returns:

```sql
> select [named_struct('a', 1, 'b', null)][1];

+-----------------------------------------------------------------------+
| make_array(named_struct(Utf8("a"),Int64(1),Utf8("b"),NULL))[Int64(1)] |
+-----------------------------------------------------------------------+
| {a: 1, b: NULL}                                                       |
+-----------------------------------------------------------------------+
1 row(s) fetched. 
Elapsed 0.007 seconds.
```

This is correct. But my previous fix (commit: d9a699ff7b787bf75da8415431fccf8cd676fa3d) returned `NULL`. So to fix the behavior while also handling out of bounds indexes, I've had to add this handler.

comphead commented on datafusion/functions-nested/src/extract.rs:
```diff
@@ -213,6 +213,33 @@ fn array_element_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
+    i64: TryInto<O>,
+{
+    let index: O = index.try_into().map_err(|_| {
+        DataFusionError::Execution(format!("array_element got invalid index: {index}"))
```
```suggestion
        exec_err!("array_element got invalid index: {index}")
```

krishvishal replied:
Done.

krishvishal commented on datafusion/functions-nested/src/extract.rs:
```diff
@@ -225,44 +252,37 @@ where
 
-        // array is null
-        if len == O::usize_as(0) {
+        if array.is_null(row_index) || len == O::usize_as(0) {
```
From my understanding

`is_null()` checks if an element at an index is `NULL`. 

2nd condition checks if the element is an empty list. 

They are checking different things right?

krishvishal replied:
Oh, I understand what you mean.

krishvishal replied:
Done.

krishvishal commented:
@comphead I've pushed additional fixes you've asked. I think the PR is ready. Let me know if it is okay.
